### ISSUE TITLE
LEAK: ByteBuf.release() was not called before it's garbage-collected.

### ISSUE THREAD

**Issue by avcad** [redisson/redisson#3502]
Redisson 3.5.1

I am debugging why we are seeing ByteBuf leaks and native memory problems.
I suspect one reason could be serialization failing for some objects.

I created a simple test to run this code in a loop
Object MY_DUMMY_NON_SERIALIZABLE_OBJECT = Logger.getGlobal();
redisson.getMap("cacheMap").fastPut("value", MY_DUMMY_NON_SERIALIZABLE_OBJECT );

The code triggers: java.lang.IllegalArgumentException: java.io.NotSerializableException: java.util.logging.Logger
and looks like MarshallingCodec. Encoder.encode doesn't clean the ByteBuf  if there is an exception - and I 
don't see a way to do it myself.  

I enabled -Dio.netty.leakDetectionLevel=paranoid to get this output:

```
SEVERE: LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
	io.netty.buffer.AdvancedLeakAwareByteBuf.writeBytes(AdvancedLeakAwareByteBuf.java:616)
	org.redisson.codec.MarshallingCodec$ByteOutputWrapper.write(MarshallingCodec.java:131)
	org.jboss.marshalling.UTFUtils.writeUTFBytes(UTFUtils.java:131)
	org.jboss.marshalling.river.RiverMarshaller.writeKnownObject(RiverMarshaller.java:397)
	org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:217)
	org.jboss.marshalling.AbstractObjectOutput.writeObject(AbstractObjectOutput.java:58)
	org.jboss.marshalling.AbstractMarshaller.writeObject(AbstractMarshaller.java:111)
	org.redisson.codec.MarshallingCodec$4.encode(MarshallingCodec.java:170)
	org.redisson.command.CommandAsyncService.encodeMapKey(CommandAsyncService.java:725)
	org.redisson.RedissonObject.encodeMapKey(RedissonObject.java:301)
	org.redisson.RedissonMap.fastPutOperationAsync(RedissonMap.java:1375)
	org.redisson.RedissonMap.fastPutAsync(RedissonMap.java:1365)
	org.redisson.RedissonMap.fastPut(RedissonMap.java:1380)
	adhoc.testing.RedissonMemoryTest.main(RedissonMemoryTest.java:53)
#2:
	io.netty.buffer.AdvancedLeakAwareByteBuf.writeBytes(AdvancedLeakAwareByteBuf.java:616)
	org.redisson.codec.MarshallingCodec$ByteOutputWrapper.write(MarshallingCodec.java:131)
	org.jboss.marshalling.SimpleDataOutput.shallowFlush(SimpleDataOutput.java:353)
	org.jboss.marshalling.river.RiverMarshaller.writeKnownObject(RiverMarshaller.java:396)
	org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:217)
	org.jboss.marshalling.AbstractObjectOutput.writeObject(AbstractObjectOutput.java:58)
	org.jboss.marshalling.AbstractMarshaller.writeObject(AbstractMarshaller.java:111)
	org.redisson.codec.MarshallingCodec$4.encode(MarshallingCodec.java:170)
	org.redisson.command.CommandAsyncService.encodeMapKey(CommandAsyncService.java:725)
	org.redisson.RedissonObject.encodeMapKey(RedissonObject.java:301)
	org.redisson.RedissonMap.fastPutOperationAsync(RedissonMap.java:1375)
	org.redisson.RedissonMap.fastPutAsync(RedissonMap.java:1365)
	org.redisson.RedissonMap.fastPut(RedissonMap.java:1380)
	adhoc.testing.RedissonMemoryTest.main(RedissonMemoryTest.java:53)
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:385)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:173)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:107)
	org.redisson.codec.MarshallingCodec$4.encode(MarshallingCodec.java:165)
	org.redisson.command.CommandAsyncService.encodeMapKey(CommandAsyncService.java:725)
	org.redisson.RedissonObject.encodeMapKey(RedissonObject.java:301)
	org.redisson.RedissonMap.fastPutOperationAsync(RedissonMap.java:1375)
	org.redisson.RedissonMap.fastPutAsync(RedissonMap.java:1365)
	org.redisson.RedissonMap.fastPut(RedissonMap.java:1380)
	adhoc.testing.RedissonMemoryTest.main(RedissonMemoryTest.java:53)
```

If I can't get the native memory issues sorted out, I need to see if this can help:
-Dio.netty.noPreferDirect=true  -Dio.netty.allocator.type=unpooled

**Comment by avcad** [comment]
Another scenario: If I create a batch and add a few values with no error, but then I add one more and that fails with NotSerializableException, how can I cancel the whole batch before execute() and free up all the native buffers?

**Comment by avcad** [comment]
This might be for a separate issue. It's minor.. but I noticed that it's hard for the calling code to tell 
if the error is due to de-serialization. see these examples:

org.redisson.client.RedisException: Unexpected exception while processing command cause: net.jpountz.lz4.LZ4Exception: Error decoding offset 10 of input buffer
com.fasterxml.jackson.core.JsonParseException: Illegal character ((CTRL-CHAR, code 4)): only regular white space (\r, \n,\t) is allowed between tokens
org.redisson.client.RedisException: Unexpected exception while processing command cause: java.io.IOException: Unsupported protocol version 51

It can be nice for the calling code to be able to tell if the issue is de-serialization vs something else. I noticed that sometimes
some entries don't fail to serialize but they do fail to de-serialize. 

As you are updating this code, could the encode/decode code change for example to wrap the exceptions in to a RedissonEncoding/DecodingException that extends IOException ? This way the calling code will have a generic way to tell if the error is with serialization, regardless of the encoder used.

**Comment by mrniko** [comment]
> Even if ByteBuf is released in a finally block, we could have one, for example, for the key that didn't fail
and remains allocated
return commandExecutor.writeAsync(name, codec, RedisCommands.HSET, name, encodeMapKey(key), encodeMapValue(value));

I agree this case should be handled.

> If I create a batch and add a few values with no error, but then I add one more and that fails with NotSerializableException, how can I cancel the whole batch before execute() and free up all the native buffers?

There is no such function to release buffers. Can you create separate issue?

> As you are updating this code, could the encode/decode code change for example to wrap the exceptions in to a RedissonEncoding/DecodingException that extends IOException ?

I can't extend these errors from IOException since it requires to wrap every command in try/catch block.

**Comment by boot-vue** [comment]
I hava the problem also.

**Comment by mrniko** [comment]
@boot-vue 

Which exactly? 

> The code triggers: java.lang.IllegalArgumentException: java.io.NotSerializableException: java.util.logging.Logger
and looks like MarshallingCodec. Encoder.encode doesn't clean the ByteBuf if there is an exception - and I
don't see a way to do it myself.

This has been already fixed

**Comment by boot-vue** [comment]
> @boot-vue
> 
> Which exactly?
> 
> > The code triggers: java.lang.IllegalArgumentException: java.io.NotSerializableException: java.util.logging.Logger
> > and looks like MarshallingCodec. Encoder.encode doesn't clean the ByteBuf if there is an exception - and I
> > don't see a way to do it myself.
> 
> This has been already fixed


Ignore my reply for now. 

I've got this problem , but I'm not sure if it's Redisson's bug.  Let me do it later

**Comment by boot-vue** [comment]
after many tests and  Jmeter  stress tests , I found the error not  reappear, I don't change anything.  😂😂😂

`redisson-spring-boot-starter : 3.15.3`

or maybe it caused by others.

**Comment by mrniko** [comment]
> If I create a batch and add a few values with no error, but then I add one more and that fails with NotSerializableException, how can I cancel the whole batch before execute() and free up all the native buffers?

Can you provide code example?

**Comment by anabright** [comment]
Hi, do you guys have an estimate of when this fix will be released?

**Comment by mrniko** [comment]
@anabright 

Please describe your case. RBatch.discard() method was introduced.

**Comment by anabright** [comment]
Sorry, I meant the fix to this problem 

>The code triggers: java.lang.IllegalArgumentException: java.io.NotSerializableException: java.util.logging.Logger
>and looks like MarshallingCodec. Encoder.encode doesn't clean the ByteBuf if there is an exception - and I
>don't see a way to do it myself.

I just realised it's already included in the latest release, I thought it was still unreleased because this issue's still marked as open. Thanks!

**Comment by mrniko** [comment]
@anabright 

That fix already included in 3.15.3
